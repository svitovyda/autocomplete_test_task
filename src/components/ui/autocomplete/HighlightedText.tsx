import type { QueryType } from '../../../utils/textSearch';
import React from 'react';
import styled from 'styled-components';

interface HighlightedTextProps {
  text: string;
  query: QueryType;
  caseSensitive?: boolean;
}

const Highlight = styled.span`
  background-color: yellow; /* or any other highlight color */
  font-weight: bold;
`;

export const HighlightedText: React.FC<HighlightedTextProps> = React.memo(
  ({ text, query, caseSensitive = false }: HighlightedTextProps) => {
    if (!query) return <>{text}</>;

    const regex = new RegExp(`(${query})`, caseSensitive ? 'g' : 'gi');
    const parts = text.split(regex);

    return (
      <>
        {parts.map((part, index) =>
          regex.test(part) ? <Highlight key={index}>{part}</Highlight> : part
        )}
      </>
    );
  }
);

HighlightedText.displayName = 'HighlightedText';
