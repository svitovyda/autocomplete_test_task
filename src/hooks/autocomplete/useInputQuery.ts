import type { QueryType } from '../../utils/textSearch';
import debounce from 'debounce';
import * as React from 'react';

type InputQueryActionType =
  | 'INPUT_CHANGED' // user edits the value
  | 'QUERY_DEBOUNCED'; // debounced value dispatched

export interface InputQueryState {
  input: string;
  query: QueryType;
}

type InputQueryPayload = Partial<InputQueryState>;

interface InputQueryAction {
  type: InputQueryActionType;
  payload: InputQueryPayload;
}

export const initialInputQueryState: InputQueryState = {
  input: '',
  query: undefined,
};

const inputStateReducer = (state: InputQueryState, action: InputQueryAction): InputQueryState => {
  switch (action.type) {
    case 'INPUT_CHANGED':
      if (state.input !== action.payload.input) {
        return { ...state, input: action.payload.input || '' };
      }
      return state;

    case 'QUERY_DEBOUNCED':
      if (state.query !== action.payload.query) {
        return { ...state, query: action.payload.query, input: action.payload.input || '' };
      }
      return state;

    default:
      throw new Error();
  }
};

export const useInputQuery = (
  debounceInterval: number = 600,
  minAcceptableLength: number = 2
): [
  InputQueryState,
  React.Dispatch<React.SetStateAction<string>>,
  // React.Dispatch<React.SetStateAction<QueryType>>,
] => {
  const [input, setInput] = React.useState<string>(''); // current state of the input

  const [state, dispatch] = React.useReducer(inputStateReducer, initialInputQueryState);

  // eslint doesn't know how to work with debounce
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const updateSearchQuery = React.useCallback(
    debounce((value: string) => {
      dispatch({
        type: 'QUERY_DEBOUNCED',
        payload: { query: value, input: value },
      });
    }, debounceInterval),
    [dispatch]
  );

  React.useEffect(() => {
    if (!input || input.length >= minAcceptableLength) {
      updateSearchQuery(input);
    }
    dispatch({ type: 'INPUT_CHANGED', payload: { input } });
  }, [input, minAcceptableLength, updateSearchQuery]);

  return [state, setInput];
};
