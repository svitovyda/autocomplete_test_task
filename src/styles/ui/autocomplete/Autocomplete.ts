import styled from 'styled-components';

export const AutocompleteInput = styled.input`
  display: inline-block;
  padding: 4px 2px;
  cursor: text;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #aaa;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin: 0;
  padding: 8px;
  outline-width: 0;
  color: inherit;
  ::placeholder {
    color: #aaa;
    font-size: 16px;
  }
`;
