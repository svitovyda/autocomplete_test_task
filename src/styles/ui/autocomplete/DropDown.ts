import styled, { css } from 'styled-components';

export const DropDownContainer = styled.div`
  width: 100px;
  margin: 0;
  padding: 0;
  display: block;
  height: 0;
`;

export const DropdownList = styled.ul`
  position: absolute;
  box-sizing: border-box;
  list-style: 'none';
  width: 200px;
  min-width: fit-content;
  max-height: 300px;
  overflow-y: auto;
  margin: 0;
  padding: 0;
  list-style-type: none;
  background-color: white;
  border: 1px solid #aaa;
  border-radius: 3px;
  box-shadow: 3px 3px 5px 0 #999;
  z-index: 100000;
  scrollbar-width: none;
`;

css`
  option {
    box-sizing: border-box;
    margin: 0;
    padding: 4px 2px;
    font-size: 18px;
    font-weight: 400;
    width: auto;
    color: #543729;
  }
`;

export const DropdownItem = styled.li`
  padding: 4px 10px;
  font-size: 16px;
  font-weight: 400;
  width: auto;
  cursor: pointer;
  &:hover {
    background-color: #d0d0d0;
    border: '1px solid #ccc';
  }
`;
