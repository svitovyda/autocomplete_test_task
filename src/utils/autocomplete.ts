import type { DataArrayType } from '../hooks/autocomplete/useDataShow';
import type { DictionaryItem, ItemID } from '../models/Autocomplete';

export const eq = <T>(a: T): T => a;

export const defaultLabelId = <T>(item: T): string => {
  return (item as any).toString();
};

export const convertToShow = <T>(
  items: DataArrayType<T>,
  dataId: (item: T) => ItemID = defaultLabelId,
  dataLabel: (item: T) => string = defaultLabelId,
  max: number = 5
): DictionaryItem<T>[] =>
  items
    .slice(0, max)
    .map((item) => ({ itemData: item, itemToShow: { id: dataId(item), label: dataLabel(item) } }));
