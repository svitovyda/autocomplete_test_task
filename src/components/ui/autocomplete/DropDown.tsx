/* eslint-disable react/prop-types */
import type { DataArrayType } from '../../../hooks/autocomplete/useDataShow';
import type { DictionaryItem } from '../../../models/Autocomplete';
import {
  DropDownContainer,
  DropdownItem,
  DropdownList,
} from '../../../styles/ui/autocomplete/DropDown';
import type { QueryType } from '../../../utils/textSearch';
import { HighlightedText } from './HighlightedText';
import _ from 'lodash';
import React from 'react';

interface DropdownProps<T> {
  items: DataArrayType<DictionaryItem<T>>;
  query?: QueryType;
  caseSensitive?: boolean;
  onSelectItem: (item: DictionaryItem<T>) => void;
}

export const DropDown = React.memo(
  React.forwardRef<HTMLDivElement, DropdownProps<any>>(
    <T,>(
      { items, query, caseSensitive = false, onSelectItem }: DropdownProps<T>,
      ref: React.ForwardedRef<HTMLDivElement>
    ) => {
      return (
        <DropDownContainer ref={ref}>
          <DropdownList>
            {items.map((item) => (
              <DropdownItem key={item.itemToShow.id} onClick={() => onSelectItem(item)}>
                <HighlightedText
                  text={item.itemToShow.label}
                  query={query}
                  caseSensitive={caseSensitive}
                />
              </DropdownItem>
            ))}
          </DropdownList>
        </DropDownContainer>
      );
    }
  ),
  _.isEqual
);

DropDown.displayName = 'DropDown';
