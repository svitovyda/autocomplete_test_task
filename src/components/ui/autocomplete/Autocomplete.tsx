import { type DataArrayType, useDataShow } from '../../../hooks/autocomplete/useDataShow';
import { useInputQuery } from '../../../hooks/autocomplete/useInputQuery';
import { useClickOutside } from '../../../hooks/useClickOutside';
import type { DictionaryItem } from '../../../models/Autocomplete';
import { AutocompleteInput } from '../../../styles/ui/autocomplete/Autocomplete';
import { defaultLabelId } from '../../../utils/autocomplete';
import type { QueryType } from '../../../utils/textSearch';
import { DropDown } from './DropDown';
import _ from 'lodash';
import * as React from 'react';

export type ItemID = string | number;

export interface ShowItem {
  id: ItemID;
  label: string;
}

export interface AutocompleteProps<T> {
  onQueryChanged?: (query: QueryType) => void;
  onItemSelected?: (item: T) => void;
  minAcceptableLength?: number;
  debounceInterval?: number;
  placeholder?: string;
  maxOptionsToShow?: number;
  autoCalculate?: boolean;
  data: DataArrayType<T>;
  dataLabel?: (item: T) => string;
  dataId?: (item: T) => ItemID;
  caseSensitive?: boolean;
  showDataOnEmptyInput?: boolean;
  initialInput?: string;
  loading?: boolean;
  LoadingAnimation?: React.ComponentType<React.HTMLAttributes<HTMLElement>>;
  keepDropDownOnSelect?: boolean;
}
/*
 * This component renders input with autocompletion
 * It can be in two modes:
 * - autoCalculate ("static")
 *   it gets all the data once and then as user interacts, searches through this given data
 * - "dynamic"
 *   it assumes that all the calculations with data search by query happen outside of the component,
 *   and it just renders data
 */
export const Autocomplete = React.memo(
  <T,>({
    onQueryChanged,
    onItemSelected,
    minAcceptableLength = 3,
    debounceInterval = 600,
    placeholder = '',
    maxOptionsToShow = 5,
    autoCalculate,
    data = [],
    dataLabel = defaultLabelId,
    dataId = defaultLabelId,
    caseSensitive = false,
    showDataOnEmptyInput = true,
    initialInput,
    loading,
    LoadingAnimation,
    keepDropDownOnSelect,
  }: AutocompleteProps<T>) => {
    /**
     * If Autocomplete is used to display result of the search that were recalculated outside,
     * `autoCalculate` should be set to false. Then we just always need to convert each item to
     * pair `{id, label}`, and take first `maxOptionsToShow`.
     * So `showInit` should always be `true` when `autoCalculate` is false
     *
     * And only once user clicks on the input or focuses on it, this precalculated pairs should be
     * shown
     *
     * If Autocomplete is using static data and should filter it by itself as user types,
     * `showDataOnEmptyInput` lets us to choose, what to show when user didn't enter the input:
     * no options or just first `maxOptionsToShow` options
     **/

    // const [inputQueryState, setInput, setQuery] = useInputQuery(
    const [inputQueryState, setInput] = useInputQuery(debounceInterval, minAcceptableLength);
    const prevQuery = React.useRef<QueryType>(undefined);

    const [showDropDown, setShowDropDown] = React.useState(false);

    const itemsToShow = useDataShow<T>(
      data,
      dataLabel,
      dataId,
      maxOptionsToShow,
      autoCalculate,
      caseSensitive,
      inputQueryState.query
    );

    const inputRef = React.useRef<HTMLInputElement>(null);
    const dropDownRef = React.useRef<HTMLDivElement>(null);

    React.useEffect(() => {
      if (initialInput !== undefined) {
        setInput(initialInput || '');
      }
    }, [initialInput, setInput]);

    // listens to the query debounced, to make HTTP call for example
    // it should be called no matter if query changed due to user typing or user selected item,
    // we need to make sure external search result has all the correct data from server
    React.useEffect(() => {
      if (inputQueryState.query !== prevQuery.current) {
        prevQuery.current = inputQueryState.query;
        onQueryChanged?.(inputQueryState.query);
      }
    }, [inputQueryState.query, onQueryChanged]);

    // all the calculations of dataToShow should be made by this moment,
    // only show drop-down and select all tezt in input
    const onFocusIn = React.useCallback(() => {
      if (showDataOnEmptyInput || inputQueryState.input) {
        setShowDropDown(true);
        if (inputRef.current && inputQueryState.input) {
          inputRef.current.setSelectionRange(0, inputQueryState.input.length);
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dropDownRef?.current, inputRef?.current, inputQueryState.input, itemsToShow]);

    // listens to user typing
    const handleInputChange = React.useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const { value } = e.target;
        setInput(value);
      },
      [setInput]
    );

    // if user clicks outside of this component - close drop-down
    const onClickOutside = React.useCallback(() => {
      setShowDropDown(false);
    }, []);

    useClickOutside([dropDownRef, inputRef], onClickOutside);

    // listens on user selecting the option from drop-down
    const onSelectItem = React.useCallback(
      (item: DictionaryItem<T>) => {
        onItemSelected?.(item.itemData);
        keepDropDownOnSelect && setShowDropDown(false);
        setInput(item.itemToShow.label);
      },
      [onItemSelected, setInput, keepDropDownOnSelect]
    );

    return (
      <>
        <AutocompleteInput
          ref={inputRef}
          type="text"
          value={inputQueryState.input}
          onChange={handleInputChange}
          placeholder={placeholder}
          onFocus={onFocusIn}
        />

        {showDropDown && itemsToShow.length > 0 && (
          <DropDown
            ref={dropDownRef}
            onSelectItem={onSelectItem}
            items={itemsToShow}
            query={inputQueryState.query}
            caseSensitive={caseSensitive}
            loading={loading}
            LoadingAnimation={LoadingAnimation}
          />
        )}
      </>
    );
  },
  (prevProps: AutocompleteProps<any>, nextProps: AutocompleteProps<any>) => {
    const {
      dataLabel: prevDataLabel,
      dataId: prevDataId,
      onItemSelected: onPrevItemSelected,
      onQueryChanged: onPrevQueryChanged,
      ...prevRest
    } = prevProps;
    const {
      dataLabel: nextDataLabel,
      dataId: nextDataId,
      onItemSelected: onNextItemSelected,
      onQueryChanged: onNextQueryChanged,
      ...nextRest
    } = nextProps;
    return (
      _.isEqual(prevRest, nextRest) &&
      prevDataId === nextDataId &&
      prevDataLabel === nextDataLabel &&
      onPrevItemSelected === onNextItemSelected &&
      onPrevQueryChanged === onNextQueryChanged
    );
  }
);

Autocomplete.displayName = 'Autocomplete';
