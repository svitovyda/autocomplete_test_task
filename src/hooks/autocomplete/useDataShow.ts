import type { DictionaryItem, ItemID } from '../../models/Autocomplete';
import { convertToShow, defaultLabelId } from '../../utils/autocomplete';
import { type QueryType, filterData } from '../../utils/textSearch';
import _ from 'lodash';
import * as React from 'react';

export type DataArrayType<T> = T[] | ReadonlyArray<T>;

interface PropsToCompare<T> {
  data: DataArrayType<T>;
  maxOptionsToShow: number;
  autoCalculate: boolean;
  caseSensitive: boolean;
  query: QueryType;
}

const recalculateDataShow = <T>(
  data: DataArrayType<T>,
  dataLabel: (item: T) => string,
  dataId: (item: T) => ItemID,
  maxOptionsToShow: number,
  autoCalculate: boolean,
  caseSensitive: boolean,
  query: QueryType
): DictionaryItem<T>[] => {
  if (autoCalculate && query) {
    return convertToShow<T>(
      filterData(data, query, caseSensitive, dataLabel),
      dataId,
      dataLabel,
      maxOptionsToShow
    );
  }
  return convertToShow<T>(data, dataId, dataLabel, maxOptionsToShow);
};

export const useDataShow = <T>(
  data: DataArrayType<T>,
  dataLabel: (item: T) => string = defaultLabelId,
  dataId: (item: T) => ItemID = defaultLabelId,
  maxOptionsToShow: number = 5,
  autoCalculate: boolean = false,
  caseSensitive: boolean = false, // gets used only if autoCalculate=true
  query: QueryType = undefined // gets used only if autoCalculate=true
): DictionaryItem<T>[] => {
  const prevProps = React.useRef<PropsToCompare<T>>({
    data: [],
    maxOptionsToShow,
    autoCalculate,
    caseSensitive,
    query,
  });

  const [itemsToShow, setItemsToShow] = React.useState<DictionaryItem<T>[]>([]);

  React.useEffect(() => {
    const props: PropsToCompare<T> = {
      data,
      maxOptionsToShow,
      autoCalculate,
      caseSensitive,
      query,
    };

    if (!_.isEqual(prevProps.current, props)) {
      prevProps.current = props;

      setItemsToShow(
        recalculateDataShow<T>(
          data,
          dataLabel,
          dataId,
          maxOptionsToShow,
          autoCalculate,
          caseSensitive,
          query
        )
      );
    }
  }, [query, data, dataLabel, dataId, maxOptionsToShow, autoCalculate, caseSensitive, prevProps]);

  return itemsToShow;
};
