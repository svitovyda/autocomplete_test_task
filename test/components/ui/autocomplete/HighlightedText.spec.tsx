import { HighlightedText } from '../../../../src/components/ui/autocomplete/HighlightedText';
import type { QueryType } from '../../../../src/utils/textSearch';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import * as React from 'react';

const mockString: string = 'test_Test_some_TEST_halo';
const query: QueryType = 'test';

describe('HighlightedText', () => {
  it('renders without error empty', () => {
    const component = render(<HighlightedText query="" text="" />);

    expect(component).toMatchSnapshot();
  });

  it('renders "test_Test_some_TEST_halo", empty query', () => {
    const component = render(<HighlightedText query="" text={mockString} />);

    expect(component).toMatchSnapshot();
  });

  it('renders "test_Test_some_TEST_halo", no case sensitive, query="test"', () => {
    const component = render(<HighlightedText query={query} text={mockString} />);

    expect(component).toMatchSnapshot();
  });

  it('renders "test_Test_some_TEST_halo", case sensitive, query="test"', () => {
    const component = render(<HighlightedText query={query} text={mockString} />);

    expect(component).toMatchSnapshot();
  });
});
