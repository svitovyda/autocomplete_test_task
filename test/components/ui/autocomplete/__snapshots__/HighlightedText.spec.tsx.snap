// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightedText renders "test_Test_some_TEST_halo", case sensitive, query="test" 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: yellow;
  font-weight: bold;
}

<body>
    <div>
      <span
        class="c0"
      >
        test
      </span>
      _
      <span
        class="c0"
      >
        Test
      </span>
      _some_
      <span
        class="c0"
      >
        TEST
      </span>
      _halo
    </div>
  </body>,
  "container": .c0 {
  background-color: yellow;
  font-weight: bold;
}

<div>
    <span
      class="c0"
    >
      test
    </span>
    _
    <span
      class="c0"
    >
      Test
    </span>
    _some_
    <span
      class="c0"
    >
      TEST
    </span>
    _halo
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`HighlightedText renders "test_Test_some_TEST_halo", empty query 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      test_Test_some_TEST_halo
    </div>
  </body>,
  "container": <div>
    test_Test_some_TEST_halo
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`HighlightedText renders "test_Test_some_TEST_halo", no case sensitive, query="test" 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: yellow;
  font-weight: bold;
}

<body>
    <div>
      <span
        class="c0"
      >
        test
      </span>
      _
      <span
        class="c0"
      >
        Test
      </span>
      _some_
      <span
        class="c0"
      >
        TEST
      </span>
      _halo
    </div>
  </body>,
  "container": .c0 {
  background-color: yellow;
  font-weight: bold;
}

<div>
    <span
      class="c0"
    >
      test
    </span>
    _
    <span
      class="c0"
    >
      Test
    </span>
    _some_
    <span
      class="c0"
    >
      TEST
    </span>
    _halo
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`HighlightedText renders without error empty 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
